Dockerfile

```diff
@@ -1,71 +1,89 @@
+RUN bundle exec rake assets:precompile
 
-# Cleanup to save space in the production image
-RUN rm -rf node_modules log tmp && \
-      rm -rf /usr/local/bundle/cache && \
-      rm -rf .env && \
-      find /usr/local/bundle/gems -name "*.c" -delete && \
-      find /usr/local/bundle/gems -name "*.h" -delete && \
-      find /usr/local/bundle/gems -name "*.o" -delete && \
-      find /usr/local/bundle/gems -name "*.html" -delete
+ENV PORT=8080
+EXPOSE ${PORT}
 
-# Build runtime image
-FROM ruby:3.1.4-alpine as production
+RUN adduser --system --no-create-home nonroot
 
-# The application runs from /app
-WORKDIR /app
+# We expect the rake assets:precompile command to create these directories, but mkdir -p will create them if they don't already exist
+RUN mkdir -p tmp log
+RUN chown -R nonroot tmp log
+RUN chown nonroot db/schema.rb
 
-# Add postgres driver library
-# Add the timezone as it's not configured by default in Alpine
-RUN apk add --update --no-cache libpq tzdata && \
-  cp /usr/share/zoneinfo/Europe/London /etc/localtime && \
-  echo "Europe/London" > /etc/timezone
+FROM base as development
 
-# Copy files generated in the builder image
-COPY --from=builder /app /app
-COPY --from=builder /usr/local/bundle/ /usr/local/bundle/
+RUN bundle config set without ""
+RUN bundle install --jobs=4 --no-binstubs --no-cache
 
-ARG GIT_COMMIT_SHA="UNKNOWN"
-ENV GIT_COMMIT_SHA=${GIT_COMMIT_SHA}
-RUN echo ${GIT_COMMIT_SHA} > ./GIT_COMMIT_SHA
+# Install gecko driver for Capybara tests
+RUN apk add firefox
+RUN wget https://github.com/mozilla/geckodriver/releases/download/v0.31.0/geckodriver-v0.31.0-linux64.tar.gz \
+    && tar -xvzf geckodriver-v0.31.0-linux64.tar.gz \
+    && rm geckodriver-v0.31.0-linux64.tar.gz \
+    && chmod +x geckodriver \
+    && mv geckodriver /usr/local/bin/
 
-ENV PORT=8080
+USER nonroot
 
-EXPOSE ${PORT}
+CMD bundle exec rails s -e ${RAILS_ENV} -p ${PORT} --binding=0.0.0.0
+
+FROM base as staging
+
+RUN bundle config set without "development"
+RUN bundle install --jobs=4 --no-binstubs --no-cache
+
+# Cleanup to reduce image size
+RUN rm -rf node_modules && \
+    rm -rf /usr/local/bundle/cache && \
+    find /usr/local/bundle/gems -name "*.c" -delete && \
+    find /usr/local/bundle/gems -name "*.h" -delete && \
+    find /usr/local/bundle/gems -name "*.o" -delete && \
+    find /usr/local/bundle/gems -name "*.html" -delete && \
+    rm -rf .env  && \
+    rm -rf app/frontend
+
+USER nonroot
+
+CMD bundle exec rails s -e ${RAILS_ENV} -p ${PORT} --binding=0.0.0.0
+
+FROM base as production
+
+# Cleanup to reduce image size
+RUN rm -rf node_modules && \
+    rm -rf /usr/local/bundle/cache && \
+    find /usr/local/bundle/gems -name "*.c" -delete && \
+    find /usr/local/bundle/gems -name "*.h" -delete && \
+    find /usr/local/bundle/gems -name "*.o" -delete && \
+    find /usr/local/bundle/gems -name "*.html" -delete && \
+    rm -rf .env  && \
+    rm -rf app/frontend
```

Comment:
Note that deleting files will only reduce image size if you squash the layers (docker build --squash) but that's still an experimental feature.

Because otherwise the file still exists in the previous layer.

Reply:
We don't think we want to use `--squash` as it's an experimental feature and we agree this isn't actually reducing the image size, so have opted to remove this. Do you think there is an advantage to keeping it to reduce attack surface?

Reply:
Fair enough. I don't think those files around really affects attack surface. 

If you did want to have a slimmer final image, the way to do it would be to have your final stage start from the Ruby image, not "base" and use `COPY --from` to take precompiled files from the base stage. But like I say, I don't think you need to bother if image size is not an issue.

Reply:
We have removed this section now - agree it's not worth switching to use `COPY --from`.